---
title: "BIMS8382 Syllabus"
author: "Introduction to Data Science for Biomedical Scientists"
output: 
  html_document
---

## General Information

This is the course content for students enrolled in BIMS8382 at The University of Virginia.

### Logistics

**_Instructors:_**

**David N Martin**

- Email: dnm5ca@virginia.edu

- GitHub: [github.com/dnm5ca](https://github.com/dnm5ca)

<small>David is the Clinical Data Research Specialist at the Health Sciences Library where he consults with clinicians, nurses, and students on data analysis and programming in a number of languages. With Marieke, he co-teaches the BIMS8380 statistics course. He has his PhD from the Curry School of Education in Educational Research, Statistics and Evaluation.</small>

**Marieke K Jones**

- Email: marieke@virginia.edu

- GitHub: [github.com/mariekekjones](https://github.com/mariekekjones)

<small>Marieke is the Research Data Specialist in the Health Sciences Library. She teaches open enrollment workshops in R, consults with researchers about statistics and data analysis, and co-teaches the BIMS8380 statistics course. Her PhD is in Environmental Science and Policy and she conducted her dissertation research on reproductive physiology, endocrinology, and metabolomics.</small>

**_Where:_** This course will be taught synchronously online using videoconferencing tools, this course website, UVA Collab, and email

**_When:_**  
Spring 2021 Module S2 
April 1 – May 11 <small>_(No exam -- final class period will be held May 11)_</small>  
9:00 - 10:30am Tuesdays and Thursdays  


### About this course

This course introduces methods, tools, and software for reproducibly managing, manipulating, analyzing, and visualizing large-scale biomedical data. Specifically, the course introduces the R statistical computing environment and packages for manipulating and visualizing high-dimensional data, covers strategies for reproducible research, and culminates with analyses of real biomedical data using R and Bioconductor packages.

**This is not a _"Tool X"_ or _"Software Y"_ class.** We want you to take away from this series the ability to use an extremely powerful scientific computing environment (R) to do many of the things that you’ll do _across study designs and disciplines_ -- managing, manipulating, visualizing, and analyzing large, sometimes high-dimensional data. Whether that data is gene expression data from yeast, microbial genomics data from _B. pertussis_, public health data from [Gapminder](http://www.gapminder.org/), RNA-seq data from humans, influenza outbreak data, movie preference trends from Netflix, or truck routing data from FedEx, you’ll need the same computational know-how and data literacy to do the same kinds of basic tasks in each. We might show you how to use specific tools here and there, but these are not important -- you probably won't be using the same specific software or methods 10 years from now, but you'll still use the same underlying data and computational foundation. **That** is the point of this series -- to arm you with a basic foundation, and more importantly, to **enable you to figure out how to use _this tool_ or _that tool_ on your own**, when you need to.

**_This is not a statistics class._** There are short lessons on essential statistics using R and regression in R but these lessons offer neither a comprehensive background on underlying theory nor in-depth coverage of implementation strategies using R. Some general knowledge of statistics and study design is helpful, but isn’t required for this course.


## [Setup](setup.html)

**Prior to the first class session**:
You should set aside a couple hours to download, install, and test all the software needed for the course. All the software we're using in class is open-source and freely available online. This setup must be completed _prior to class_, as we will not have much time for troubleshooting software installation issues during class. [Email us](people.html) if you're having difficulty.

Because of this course's format as an online module, here we provide best options for following along during class sessions. Most of the course consists of live coding, so the challenge will be how to simultaneously view the instructor's screen and your screen given that the RStudio window is large.

1. _Best option_ Dual monitors side-by-side
2. Two computers side-by-side
3. One computer (you) and one tv (instructor)
4. One computer (you) and one tablet (instructor)
5. One computer (you & instructor) and PDF of printed codes

## Course Schedule

### April 1: Intro to R

This novice-level introduction is directed toward life scientists with little to no experience with statistical computing or bioinformatics. This interactive introduction will introduce the R statistical computing environment. The first part of this session will demonstrate very basic functionality in R, including functions, vectors, creating variables, getting help, and reading/writing files. 

### April 6: RMarkdown and dplyr

This session covers some of the barriers to reproducible research and how to start to address some of those problems. We will cover using RMarkdown and RStudio to combine text and executable code to create readable and shareable reports. In the process of writing RMarkdown documents we will also explore the _dplyr_ package. The _dplyr_ package in R will help us to effectively manipulate and conditionally compute summary statistics over subsets of a "big" dataset containing many observations.

### April 8: Data Visualization with ggplot2 

This session will cover fundamental concepts for creating effective data visualization and will introduce tools and techniques for visualizing large, high-dimensional data using R. We will review fundamental concepts for visually displaying quantitative information, such as using series of small multiples, avoiding "chart-junk," and maximizing the data-ink ratio. After briefly covering data visualization using base R graphics, we will introduce the _ggplot2_ package for advanced high-dimensional visualization. We will cover the grammar of graphics (geoms, aesthetics, stats, and faceting), and using ggplot2 to create plots layer-by-layer. Upon completing this lesson, students will be able to use R to explore a high-dimensional dataset by faceting and scaling arbitrarily complex plots in small multiples.

### April 13: Exploratory Data Analysis

This session will introduce the idea of “tidy data”, and will introduce R packages that enable data manipulation and plotting to explore a dataset and understand what steps must be taken to clean the data.

### April 15: Advanced Exploratory Analysis

Riffing off of Tolstoy, Hadley Wickham, the writer of many of the R packages we will use, says "Like families, tidy datasets are all alike but every messy dataset is messy in its own way". Upon completing this lesson, students will be able to use the _tidyr_, _stringr_, _dplyr_, and _ggplot2_ packages in R to clean and explore their data.

### April 20: TidyTuesday Challenge 1

Each week RStudio releases a new dataset in a TidyTuesday challenge so we will each load the dataset, manipulate it as needed, and create a data visualization that tells a story. At the end of the session we will share our viz with each other and look at some other coders' contributions on Twitter to get some fresh ideas.

### April 22: Essential Statistics

This workshop will provide hands-on instruction and exercises covering basic statistical analyses in R. This will cover descriptive statistics, exploratory data analysis, t-tests, ANOVA, linear models, and methods for discrete variable analysis. We will cover hypothesis testing and assessing the assumptions of the tests.

### April 27: Linear Regression

This session will focus solely on conducting regression analyses in R. It will begin with an overview of regression analysis and then move on to specific techniques which include: linear models, logistic regression, and multiple regression. We will also focus on methods to check model assumptions and to select the "best" model.

### April 29: Multiple Regression

Here we will expand our knowledge of regression models by considering models with several predictors and learning how to choose the "best fit" model.

### May 4: Logistic Regression and Predictive Analytics

For this session we will learn to model binary outcomes and utilize data science principles to create predictive models. We will follow the train, test, validate method to develop the best predictive model.

### May 6: Dimension Reduction

In this session we will investigate techniques for inspecting, visualizing, and reducing dimensions for high-dimensional data. We will examine techniques like PCA, clustering, and heatmaps that are useful for any number of large datasets.

### May 11: Intro to Single cell RNAseq and High Performance Computing (guest instructors from Research Computing)

In this session, Will Rosenow and Gladys Andino Bautista will help us explore how to perform single cell RNAseq on Rivanna, UVA's high performance computing cluster. They will demonstrate how to log onto Rivanna, setting up an R environment for running R code, and submitting jobs to run in the background. Then they will lead us through an analysis that requires the computing power of the HPC, single cell RNAseq using the Seurat package in R.

### Student Presentations

So much of learning biomedical datascience will be up to you. You each are studying a different topic with varying computational needs. As such, we would like you to choose a package we did not cover at length in class and create a vignette demoing for the class what the package is for and a few of its functions. You will turn in a rendered PDF of your vignette, provide a 10min recorded presentation with a partner, and critique presentations of your peers. This assignment is worth 20% of your course grade. As an example of a vignette, please see UVA Library's Clay Ford's vignette of the [pwr package](https://cran.r-project.org/web/packages/pwr/vignettes/pwr-vignette.html)

## Homework

Each week’s lesson will be followed by an independent homework assignment that will be posted to our course’s Collab site. The HW will be due the following Monday at 11:59pm. Homework assignments should be completed in RMarkdown and should be submitted as a single PDF file to our course’s Collab site. In total there will be 5 homework assignments due April 12, April 19, April 26, May 3, and May 10.

## Grading Policy

* 80% Homeworks -- each worth 16%
* 20% Presentation and vignette

## FAQ 

### What are the pre-requisites?

_There are none!_ This course does not assume any knowledge of programming or using a command-line interface, but if you've have had any experience here, the content won't come as so much of a shock. But _don't panic._ Command-line interfaces and programming languages like R are _incredibly powerful_ and will be utterly transformative on your research. There's a learning curve, and it's near-vertical in the beginning, but it's surmountable and the payoff is worth it! 

### Do I need a laptop?

**YES.** You must have access to a computer on which you can install software. The class will be primarily live coding with activities for you to independently code. As such, ideally you would have 2 monitors allowing you to follow along with the instructor in one screen while you code in a second. Please follow the [setup instructions](setup.html) prior to the workshop.

### Where can I get more help?

Glad you asked! [See here](help.html).