---
title: "BIMS8382 Syllabus"
author: "Introduction to Data Science for Biomedical Scientists"
output: 
  html_document
---

## General Information

### Logistics

**_Instructors:_**

**David N Martin**
- Email: dnm5ca@virginia.edu
- GitHub: [github.com/dnm5ca](https://github.com/dnm5ca)
<small>David is the Clinical Data Research Specialist at the Health Sciences Library where he consults with clinicians, nurses, and students on data analysis and programming in a number of languages. He is also a PhD candidate in the Curry School of Education researching the impacts of the No Child Left Behind policy.</small>

**Marieke K Jones**
- Email: marieke@virginia.edu
- GitHub: [github.com/mariekekjones](https://github.com/mariekekjones)
<small>Marieke is the Research Data Specialist in the Health Sciences Library. She teaches open enrollment workshops in R, consults with researchers about statistics and data analysis, and teaches the BIMS8380 statistics course. Her PhD is in Environmental Science and Policy and she conducted her dissertation research on reproductive physiology and endocrinology.</small>

**_Where:_** BIMS Education Center (McKim Hall 1023)

**_When:_**  
Spring 2019 Module S1  
February 11 – March 25 <small>_(No exam -- final class period will be held March 25.)_</small>  
2:00pm - 5:00pm  


### About this course

This course introduces methods, tools, and software for reproducibly managing, manipulating, analyzing, and visualizing large-scale biomedical data. Specifically, the course introduces the R statistical computing environment and packages for manipulating and visualizing high-dimensional data, covers strategies for reproducible research, and culminates with analyses of real biomedical data using R and Bioconductor packages.

**This is not a _"Tool X"_ or _"Software Y"_ class.** We want you to take away from this series the ability to use an extremely powerful scientific computing environment (R) to do many of the things that you’ll do _across study designs and disciplines_ -- managing, manipulating, visualizing, and analyzing large, sometimes high-dimensional data. Whether that data is gene expression data from yeast, microbial genomics data from _B. pertussis_, public health data fromm [Gapminder](http://www.gapminder.org/), RNA-seq data from humans, influenza outbreak data, movie preference trends from Netflix, or truck routing data from FedEx, you’ll need the same computational know-how and data literacy to do the same kinds of basic tasks in each. We might show you how to use specific tools here and there, but these are not important -- you probably won't be using the same specific software or methods 10 years from now, but you'll still use the same underlying data and computational foundation. **That** is the point of this series -- to arm you with a basic foundation, and more importantly, to **enable you to figure out how to use _this tool_ or _that tool_ on your own**, when you need to.

**_This is not a statistics class._** There are short lessons on essential statistics using R and regression in R but these lessons offer neither a comprehensive background on underlying theory nor in-depth coverage of implementation strategies using R. Some general knowledge of statistics and study design is helpful, but isn’t required for this course.


## [Setup](setup.html)

**Prior to the first class session**:
You should set aside a couple hours to download, install, and test all the software needed for the course. All the software we're using in class is open-source and freely available online. This setup must be completed _prior to class_, as we will not have much time for troubleshooting software installation issues during class. [Email us](people.html) if you're having difficulty.

After installing and testing the software we'll be using, please download the data as instructed, and review any required reading _prior to class_.


## Course Schedule

_(Subject to change)_

### Week 1: Intro to R (David)

This novice-level introduction is directed toward life scientists with little to no experience with statistical computing or bioinformatics. This interactive introduction will introduce the R statistical computing environment. The first part of this session will demonstrate very basic functionality in R, including functions, functions, vectors, creating variables, getting help, filtering, data frames, plotting, and reading/writing files. The final part of the session covers some of the barriers to reproducible research and how to start to address some of those problems. We will cover using RMarkdown and RStudio to combine text and executable code to create readable and shareable reports.

### Week 2: Dealing with Data (David)

Data analysis involves a large amount of janitor work – munging and cleaning data to facilitate downstream data analysis. This session assumes a basic familiarity with R and covers tools and techniques for data cleaning and manipulation. It will cover "tidy data", and will introduce R packages that enable data manipulation and analysis using split-apply-combine strategies. Upon completing this lesson, students will be able to use the _tidyr_, _stringr_ and _dplyr_ package in R to effectively manipulate and conditionally compute summary statistics over subsets of a "big" dataset containing many observations.

### Week 3: Data Visualization with ggplot2 (David)

This session will cover fundamental concepts for creating effective data visualization and will introduce tools and techniques for visualizing large, high-dimensional data using R. We will review fundamental concepts for visually displaying quantitative information, such as using series of small multiples, avoiding "chart-junk," and maximizing the data-ink ratio. After briefly covering data visualization using base R graphics, we will introduce the _ggplot2_ package for advanced high-dimensional visualization. We will cover the grammar of graphics (geoms, aesthetics, stats, and faceting), and using ggplot2 to create plots layer-by-layer. Upon completing this lesson, students will be able to use R to explore a high-dimensional dataset by faceting and scaling arbitrarily complex plots in small multiples.

### Week 4: Essential Statistics (Marieke)

This workshop will provide hands-on instruction and exercises covering basic statistical analyses in R. This will cover descriptive statistics, exploratory data analysis, t-tests, ANOVA, linear models, and methods for discrete variable analysis. We will cover hypothesis testing and assessing the assumptions of the tests.

### Week 5: Regression (Marieke)

This session will focus solely on conducting regression analyses in R. It will begin with an overview of regression analysis and then move on to specific techniques which include: linear models, logistic regression, multiple regression, mixed effects modeling, and others. We will also focus on methods to check model assumptions and to select the "best" model.

### Week 6: RNASeq (Guest)

Description TBD


### Week 7: Predictive Modeling (Marieke)

This session will provide hands-on instruction for using machine learning algorithms for prediction. We will cover data cleaning, feature extraction, imputation, and using a variety of models to try to predict disease outcome. We will use resampling strategies to assess the performance of predictive modeling procedures such as Random Forest, elastic net regularized regression (LASSO), and k-nearest neighbors.

## Homework

Each week's lesson will be followed by an independent homework assignment that will be posted to our course's Collab site. The HW will be due the following Sunday at 11:59pm. Homework assignments should be completed in RMarkdown and should be submitted as a single PDF file to our course's Collab site

## Grading Policy
This course will be graded as Pass / Fail. As long as you make a good faith effort to learn the material and submit high-quality homeworks on time, you will pass this course.

## FAQ

### What are the pre-requisites?

_There are none!_ This course does not assume any knowledge of programming or using a command-line interface, but if you've have had any experience here, the content won't come as so much of a shock. But _don't panic._ Command-line interfaces and programming languages like R are _incredibly powerful_ and will be utterly transformative on your research. There's a learning curve, and it's near-vertical in the beginning, but it's surmountable and the payoff is worth it! 


### Do I need a laptop?

**YES.** You must have access to a computer on which you can install software. The class will be a mix of lecture, discussion, but primarily live coding. You must bring your laptop to the course every day. Bring your charging cable also. Please follow the [setup instructions](setup.html) prior to the workshop.

### Where can I get more help?

Glad you asked! [See here](help.html).