<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Hugo 0.54.0" />


<title>BIMS 8382</title>
<meta property="og:title" content="BIMS 8382">
<meta property="og:type" content="website">


<link href='/images/favicon.ico' rel='icon' type='image/x-icon'/>



<meta property="description" content="BIMS 8382.">


<link href="/index.xml" rel="alternate" type="application/rss+xml" title="BIMS 8382" />

<link rel="stylesheet" href="/css/fonts.css" media="all">
<link rel="stylesheet" href="/css/main.css" media="all">


  </head>
  <body>
    <div class="wrapper">
      <header class="header">
        <nav class="nav">
  <a href="/" class="nav-logo">
    <img src="/images/logo.png"
         width="50"
         height="75"
         alt="Logo">
  </a>

  <ul class="nav-links">
    
   <li><a href="/01_IntroR/Empty.html">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Apr 1</a></li>
   <li><a href="/02_Rmd/Empty.html">Apr 6</a></li>
   <li><a href="/03_EDA/Empty.html">Apr 8</a></li>
   <li><a href="/04_DataViz/Empty.html">Apr 13</a></li>
   <li><a href="/05_Cleaning/Empty.html">Apr 15</a></li>
   <li><a href="/07_Stats/Empty.html">Apr 20</a></li>
   </ul>
   <ul class="nav-links">
     
   <li><a href="/09_Regression/Empty.html">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Apr 22</a></li>
   <li><a href="/09A_DimRed/Empty.html">Apr 27</a></li>
   <li><a href="/10_RNA/Empty.html">Apr 29</a></li>
   <li><a href="/10_RNA/Empty.html">May 4</a></li>
   <li><a href="/10_RNA/Empty.html">May 6</a></li>
   <li><a href="/10_RNA/Empty.html">May 11</a></li>
  </ul>
</nav>

      </header>


<main class="content" role="main">

  
<title>BIMS8382 Syllabus</title>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type="text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
ul {
  text-align: left;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="fluid-row" id="header">



<h1 class="title toc-ignore">BIMS8382 Syllabus</h1>
<h4 class="author">Introduction to Data Science for Biomedical Scientists</h4>

</div>


<div id="general-information" class="section level2">
<h2>General Information</h2>
<p>This is the course content for students enrolled in BIMS8382 at The University of Virginia.</p>
<div id="logistics" class="section level3">
<h3>Logistics</h3>
<p><strong><em>Instructors:</em></strong></p>
<p><strong>David N Martin</strong></p>
<ul>
<li><p>Email: <a href="mailto:dnm5ca@virginia.edu" class="email">dnm5ca@virginia.edu</a></p></li>
<li><p>GitHub: <a href="https://github.com/dnm5ca">github.com/dnm5ca</a></p></li>
</ul>
<p><small>David is the Clinical Data Research Specialist at the Health Sciences Library where he consults with clinicians, nurses, and students on data analysis and programming in a number of languages. With Marieke, he co-teaches the BIMS8380 statistics course. He has his PhD from the Curry School of Education in Educational Research, Statistics and Evaluation.</small></p>
<p><strong>Marieke K Jones</strong></p>
<ul>
<li><p>Email: <a href="mailto:marieke@virginia.edu" class="email">marieke@virginia.edu</a></p></li>
<li><p>GitHub: <a href="https://github.com/mariekekjones">github.com/mariekekjones</a></p></li>
</ul>
<p><small>Marieke is the Research Data Specialist in the Health Sciences Library. She teaches open enrollment workshops in R, consults with researchers about statistics and data analysis, and co-teaches the BIMS8380 statistics course. Her PhD is in Environmental Science and Policy and she conducted her dissertation research on reproductive physiology and endocrinology.</small></p>
<p><strong><em>Where:</em></strong> This course will be taught synchronously online using Zoom, this course website, Collab, and email</p>
<p><strong><em>When:</em></strong><br />
Spring 2021 Module S2 Apr 1 – May 11 <small><em>(No exam – final class period will be held May 11)</em></small><br />
9:00 - 10:30am Tuesdays and Thursdays</p>
</div>
<div id="about-this-course" class="section level3">
<h3>About this course</h3>
<p>This course introduces methods, tools, and software for reproducibly managing, manipulating, analyzing, and visualizing large-scale biomedical data. Specifically, the course introduces the R statistical computing environment and packages for manipulating and visualizing high-dimensional data, covers strategies for reproducible research, and culminates with analyses of real biomedical data using R and Bioconductor packages.</p>
<p><strong>This is not a <em>“Tool X”</em> or <em>“Software Y”</em> class.</strong> We want you to take away from this series the ability to use an extremely powerful scientific computing environment (R) to do many of the things that you’ll do <em>across study designs and disciplines</em> – managing, manipulating, visualizing, and analyzing large, sometimes high-dimensional data. Whether that data is gene expression data from yeast, microbial genomics data from <em>B. pertussis</em>, public health data from <a href="http://www.gapminder.org/">Gapminder</a>, RNA-seq data from humans, influenza outbreak data, movie preference trends from Netflix, or truck routing data from FedEx, you’ll need the same computational know-how and data literacy to do the same kinds of basic tasks in each. We might show you how to use specific tools here and there, but these are not important – you probably won’t be using the same specific software or methods 10 years from now, but you’ll still use the same underlying data and computational foundation. <strong>That</strong> is the point of this series – to arm you with a basic foundation, and more importantly, to <strong>enable you to figure out how to use <em>this tool</em> or <em>that tool</em> on your own</strong>, when you need to.</p>
</div>
</div>
<div id="setup" class="section level2">
<h2><a href="setup.html">Setup</a></h2>
<p><strong>Prior to the first class session</strong>: You should set aside a couple hours to download, install, and test all the software needed for the course. All the software we’re using in class is open-source and freely available online. This setup must be completed <em>prior to class</em>, as we will not have much time for troubleshooting software installation issues during class. <a href="people.html">Email us</a> if you’re having difficulty.</p>
<p>Because of this course’s format as an online module, here we provide best options for following along during class sessions. Most of the course consists of live coding, so the challenge will be how to simultaneously view the instructor’s screen and your screen given that the RStudio window is large.</p>
<ol style="list-style-type: decimal">
<li><em>Best option</em> Dual monitors side-by-side</li>
<li>Two computers side-by-side</li>
<li>One computer (you) and one tv (instructor)</li>
<li>One computer (you) and one tablet (instructor)</li>
<li>One computer (you &amp; instructor) and PDF of printed codes</li>
</ol>
</div>
<div id="course-schedule" class="section level2">
<h2>Course Schedule</h2>
<div id="march-26-intro-to-r" class="section level3">
<h3>Apr 1: Intro to R</h3>
<p>This novice-level introduction is directed toward life scientists with little to no experience with statistical computing or bioinformatics. This interactive introduction will introduce the R statistical computing environment and RMarkdown for publishing reproducible reports.</p>
</div>
<div id="march-31-rmarkdown-and-dplyr" class="section level3">
<h3>Apr 6: RMarkdown and dplyr</h3>
<p>This session covers some of the barriers to reproducible research and how to start to address some of those problems. We will further cover using RMarkdown and RStudio to combine text and executable code to create readable and shareable reports. In the process of writing RMarkdown documents we will also explore the <em>dplyr</em> package. The <em>dplyr</em> package in R will help us to effectively manipulate and conditionally compute summary statistics over subsets of a dataset containing many observations.</p>
</div>
<div id="april-7-data-visualization-with-ggplot2-david" class="section level3">
<h3>April 8: Data Visualization with ggplot2</h3>
<p>This session will cover fundamental concepts for creating effective data visualization and will introduce tools and techniques for visualizing large, high-dimensional data using R. We will review fundamental concepts for visually displaying quantitative information, such as using series of small multiples, avoiding “chart-junk,” and maximizing the data-ink ratio. We will introduce the <em>ggplot2</em> package for advanced high-dimensional visualization. We will cover the grammar of graphics (geoms, aesthetics, stats, and faceting), and using ggplot2 to create plots layer-by-layer. Upon completing this lesson, students will be able to use R to explore a high-dimensional dataset by faceting and scaling arbitrarily complex plots in small multiples.</p>
</div>
<div id="april-2-exploratory-data-analysis-david" class="section level3">
<h3>April 13: Exploratory Data Analysis</h3>
<p>This session will introduce the idea of “tidy data”, and will introduce R packages that enable data manipulation and plotting to explore a dataset and understand what steps must be taken to clean the data.</p>
</div>
<div id="april-9-advanced-exploratory-analysis-david" class="section level3">
<h3>April 15: Advanced Exploratory Analysis</h3>
<p>Riffing off of Tolstoy, Hadley Wickham, the writer of many of the R packages we will use, says “Like families, tidy datasets are all alike but every messy dataset is messy in its own way”. Upon completing this lesson, students will be able to use the <em>tidyr</em>, <em>stringr</em>, <em>dplyr</em>, and <em>ggplot2</em> packages in R to clean and explore their data.</p>
</div>
<div id="april-14-tidytuesday-challenge-1" class="section level3">
<h3>April 20: TidyTuesday Challenge 1</h3>
<p>Each week RStudio releases a new dataset in a TidyTuesday challenge so we will each load the dataset, manipulate it as needed, and create a data visualization that tells a story. At the end of the session we will share our viz with each other and look at some other coders’ contributions on Twitter to get some fresh ideas.</p>
</div>
<div id="april-16-essential-statistics-marieke" class="section level3">
<h3>April 22: Essential Statistics</h3>
<p>This workshop will provide hands-on instruction and exercises covering basic statistical analyses in R. This will cover descriptive statistics, exploratory data analysis, t-tests, ANOVA, linear models, and methods for discrete variable analysis. We will cover hypothesis testing and assessing the assumptions of the tests.</p>
</div>
<div id="april-21-regression-marieke" class="section level3">
<h3>April 27: Linear Regression</h3>
<p>This session will focus solely on conducting linear regression analyses in R. It will begin with an overview of regression analysis and then move on to include fitting the model, investigating the residuals, and visualizing the models.</p>
</div>
<div id="april-23-dimension-reduction-marieke" class="section level3">
<h3>April 29: Multiple Regression</h3>
<p>Here we will expand our knowledge of regression models by considering models with several predictors and learning how to choose the "best fit" model.</p>
</div>
<div id="april-23-dimension-reduction-marieke" class="section level3">
<h3>May 4: Logistic Regression and Predictive Analytics</h3>
<p>For this session we will learn to model binary outcomes and utilize data science principles to create predictive models. We will follow the train, test, validate method to develop the best predictive model.</p>
</div>
<div id="april-23-dimension-reduction-marieke" class="section level3">
<h3>May 6: Dimension Reduction</h3>
<p>In this session we will investigate techniques for inspecting, visualizing, and reducing dimensions for high-dimensional data. We will examine techniques like PCA, clustering, and heatmaps that are useful for any number of large datasets.</p>
</div>
<div id="april-28-intro-to-high-performance-computing-jackie-huband" class="section level3">
<h3>May 11: TBD</h3>
<p>Description TBD</p>
</div>

<div id="may-5-7-student-presentations" class="section level3">
<h3>Student Presentations</h3>
<p>So much of learning biomedical datascience will be up to you. You each are studying a different topic with varying computational needs. As such, we would like you to choose a package we did not cover at length in class and create a vignette demoing for the class what the package is for and a few of its functions. You will turn in a rendered PDF of your vignette, provide a 10min recorded presentation with a partner, and critique presentations of your peers. This assignment is worth 20% of your course grade. As an example of a vignette, please see UVA Library’s Clay Ford’s vignette of the <a href="https://cran.r-project.org/web/packages/pwr/vignettes/pwr-vignette.html">pwr package</a></p>
</div>
</div>
<div id="homework" class="section level2">
<h2>Homework</h2>
<p>Each week’s lesson will be followed by an independent homework assignment that will be posted to our course’s Collab site. The HW will be due the following Monday at 11:59pm. Homework assignments should be completed in RMarkdown and should be submitted as a single PDF file to our course’s Collab site. In total there will be 5 homework assignments due April 12, April 19, April 26, May 3, and May 10.</p>
</div>
<div id="grading-policy" class="section level2">
<h2>Grading Policy</h2>
<ul>
<li>80% Homeworks – each worth 16%</li>
<li>20% Presentation and vignette</li>
</ul>
</div>
<div id="faq" class="section level2">
<h2>FAQ</h2>
<div id="what-are-the-pre-requisites" class="section level3">
<h3>What are the pre-requisites?</h3>
<p><em>There are none!</em> This course does not assume any knowledge of programming or using a command-line interface, but if you’ve have had any experience here, the content won’t come as so much of a shock. But <em>don’t panic.</em> Command-line interfaces and programming languages like R are <em>incredibly powerful</em> and will be utterly transformative on your research. There’s a learning curve, and it’s near-vertical in the beginning, but it’s surmountable and the payoff is worth it!</p>
</div>
<div id="do-i-need-a-laptop" class="section level3">
<h3>Do I need a laptop?</h3>
<p><strong>YES.</strong> You must have access to a computer on which you can install software. The class will be primarily live coding with activities for you to independently code. As such, ideally you would have 2 monitors allowing you to follow along with the instructor in one screen while you code in a second. Please follow the <a href="setup.html">setup instructions</a> prior to the workshop.</p>
</div>
<div id="where-can-i-get-more-help" class="section level3">
<h3>Where can I get more help?</h3>
<p>Glad you asked! <a href="help.html">See here</a>.</p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
